### ISSUE TITLE
Plugin and installation manager UIs

### ISSUE THREAD

**Issue by isabela-pf** [napari/packaging#27]
I've been trying to understand what is needed of the manager UIs that have been mentioned on the roadmap and working group meetings. The following is the summary of the two main levels—plugin manager and installation managers—I've been hearing about from a functionality perspective.

## Needed functionality

The following are the actions I believe we aim to cover.

- `install`
    - Installs packages to an environment. 
    - In the installation manager, this is installing napari and its dependencies.
    - In the plugin manager, this is installing a plugin.
    - Users may need to install from remote or local sources.
    - Users may need to set channels, or explicit sources, for these installations.
- `create`
    - Create a new environment.
    - In the installation manager, this is the step prior to installing napari. In the context of the bundled app, users do not trigger this directly.
- `update`
    - Update existing packages in an environment.
    - In the installation manager, this is attempting to install the latest version of napari and its dependencies.
    - In the plugin manager, this is attempting to install the latest version of a plugin.
    - Updating has its own needs. This has more detail in #21
- `list`
    - List all packages in an environment.
    - In the installation manager, this is listing intalled napari and dependency versions.
    - In the plugin manager, this is listing all installed plugins and their versions.
- `remove`
    - Remove existing packages in an environment.
    - In the installation manager, this is uninstalling napari and its dependencies.
    - In the plugin manager, this is uninstalling a plugin.
    - In the context of the bundeled app, removal may be triggered by a user if they wish to uninstall napari or a plugin.
    - When updating a napari or plugin, users do not trigger this directly. Removal of the prior version automatically happens after the new install.

Other bonuses it sounds like we'd like to include are 
- batch actions (ie. install or remove several plugins at once)
- making environment info available via the UI 

## Current status

With the existing plugin manager UI, I think we support all these main commands. Some more advanced additions (like specifying channels, batch actions) aren't yet supported.

We don't currently have an installation manager UI, so that is where most of the missing support is if I understand correctly. The installation manager also is not as immediate priority, I think.

## What next? 

With this summary, I would like to check a few things:
- Is there anything I'm missing? Is there other functionality I need to make sure and address?
- Does any of this sound inaccurate?
- Are plugin updates available via the current plugin manager?
- Is there anything that makes the bundled app version of this UI unique?

**Comment by Czaki** [comment]
For installation, the information about conflicts should be reported to the interface. Also, other cases in which installation fails should be properly reported in the interface.

Option to create the file with frozen version of napari environment (for example, to attach as supplementary material for reproducibility) and option to recreate such environment from such file (assume that user manually installs proper napari version). My motivation is the reproducibility of science. 

> * In the plugin manager, this is listing all installed plugins and their versions.

Plugins also could have some dependency, that version may create some problems.

**Comment by isabela-pf** [comment]
@Czaki thank you for adding these! By "conflicts" do you mean any error or specifically conflicts of dependencies needed to open the viewer?

I agree that I'd also like to make it a priority to surface the environment that napari is running in for reproducibility purposes. Making it produce a file does seem like the most easily portable way of this. And thanks for calling out the plugin dependencies!

**Comment by isabela-pf** [comment]
**tl;dr:** I’m unsure exactly where the gaps in the existing UI are and where immediate design help is needed. I don't know if @jaimergp or @goanpeca have thoughts on this.

---

I mentioned in [my starting comment](https://github.com/napari/packaging/issues/27#issue-1337656158) that I think the plugin manager that already exists in the non-bundled viewer meets all the actions we need for it to be functional in bundled app form. In revisiting with fresh eyes after a few days, it makes me unsure exactly where the design help is needed. 

While there are things it sounds like we’d like to add (like batch actions, info about the environment that napari is in, dependencies for plugins), my understanding is that the first steps are  to 
1. ensure that we have what we need for the bundled app viewer to function 
2. make the experiences between existing (not idealized) plugin managers in the bundled and non-bundled viewers as similar as possible.

In that case, I want to point out where I think the above actions are possible in the existing viewer to make sure we are on the same page.

<img width="1192" alt="napari viewer plugin manager window. It shows the default plugin manager with a list of installed and available plugins taking up most of the space." src="https://user-images.githubusercontent.com/50221806/184992596-bf7808b6-a1ce-41d7-9ef9-8d3d6c415515.png">

## `install`
>    - In the plugin manager, this is installing a plugin.

This is possible via the `Install` buttons on the left of the Available plugins section, by adding a URL to the input field at the bottom of the manager, or by dropping a source file in the input field at the bottom of the manager.

>    - Users may need to install from remote or local sources.

I believe the installation options above cover both these cases. Local installation seems only possible via the drag-and-drop as of now, though.

>    - Users may need to set channels, or explicit sources, for these installations.

This isn’t something I believe is possible at the moment. Would this need to be added for the bundled app UI even though I believe it’s considered an advanced feature?

## `update`
>    - In the plugin manager, this is attempting to install the latest version of a plugin.

I’m not actually sure how this currently goes in the viewer as I have encountered a plugin update yet. At worst case, I imagine a user can uninstall the current version and reinstall the new version (since that what update does behind the scenes anyways).  

I do hope we have another option I’m just not finding documentation on, though. :laughing:

>    - Updating has its own needs. This has more detail in #21

I don’t know the status of this work since it’s moved from the napari/napari repo to this napari/packaging one. Does the design exist in the viewer? Is it missing something that we've since identified as necessary?

## `list`

>  - In the plugin manager, this is listing all installed plugins and their versions.

We currently list all installed plugins, available plugins, and their versions. This does not go too deep, as far as I’m aware. For example it  wouldn’t show the dependencies or other versions available, for example. Do we need it to be that precise in the bundled app?

## `remove`
>     - In the plugin manager, this is uninstalling a plugin.

This is possible via the `Uninstall` buttons on the left of the Installed Plugins section.

---

I hope that helps clarify where I think we are and why I think I'm not sure how to be most helpful here. Input would be greatly appreciated.

**Comment by Czaki** [comment]
for example, one package requires `SimpleITK<2` and the second one `SimpleITK>=2` (more real example here: https://github.com/napari/superqt/issues/100). Then plugin dialog should have a clear message about conflicting packages to not lead to a corrupt environment. 

Even in the conda based environment, where conda could prevent such conflict, there is no clear message why the installation failed and no warnings that there could be a problem.

**Comment by Czaki** [comment]
> I’m not actually sure how this currently goes in the viewer as I have encountered a plugin update yet

From time to time update work, Time to time, I need to uninstall the plugin and install it again. I do not have time to investigate differences and understand what happen (this is about conda based bundle).

**Comment by jaimergp** [comment]
> From time to time update work, Time to time, I need to uninstall the plugin and install it again. I do not have time to investigate differences and understand what happen (this is about conda based bundle).

This is a bug with conda/mamba, which might consider that saying `conda install plugin` is already satisfied. We need to say `conda install plugin=1.2.3` to make sure it does change things. We have a ticket for this, iirc.

**Comment by isabela-pf** [comment]
> since this seems to mostly pertain to the in-app interface, should we be moving this to napari/napari? or am I misunderstanding, and this pertains more to the constructor bundle?

This is exactly what I was trying to figure out @tlambert03! It sounds to me like this is general in-app interface stuff, so I agree it might be more accurate in napari/napari. 

Would you like me to add to [napari/napari #4937](https://github.com/napari/napari/issues/4937), or does someone with permissions want to transfer this issue across repos?

**Comment by tlambert03** [comment]
sure.  Yeah if the action-item here would need to edit `napari/_qt/dialogs/qt_plugin_dialog.py`, then it should be a napari issue.  Will transfer
